identity-service/src/routes/push_notifications.js

```diff
@@ -0,0 +1,78 @@
+const { handleResponse, successResponse, errorResponseBadRequest, errorResponseServerError } = require('../apiHelpers')
+const models = require('../models')
+const DEVICE_TYPES = new Set(['ios', 'android'])
+
+module.exports = function (app) {
+  /**
+   * Create or update mobile push notification settings
+   * POST body contains {userId, settings: {favorites, milestonesAndAchievements, reposts, announcements, followers}}
+   */
+  app.post('/push_notifications/settings', handleResponse(async (req, res, next) => {
+    const { userId, settings } = req.body
+
+    if (!userId) return errorResponseBadRequest(`Did not pass in a valid userId`)
+
+    try {
+      await models.UserNotificationMobileSettings.upsert({
+        userId,
+        ...settings
+      })
+
+      return successResponse()
+    } catch (e) {
+      req.logger.error(`Unable to create or update push notification settings for userId: ${userId}`, e)
+      return errorResponseServerError(`Unable to create or update push notification settings for userId: ${userId}, Error: ${e.message}`)
+    }
+  }))
+
+  /**
+   * Register a device token
+   * POST body contains {deviceToken: <string>, deviceType: ios/android, userId}
+   */
+  app.post('/push_notifications/device_token', handleResponse(async (req, res, next) => {
+    const { deviceToken, deviceType, userId } = req.body
+
+    if (!DEVICE_TYPES.has(deviceType)) return errorResponseBadRequest('Attempting to register an invalid deviceType')
+    if (!deviceToken || !userId) return errorResponseBadRequest('Did not pass in a valid deviceToken or userId for device token registration')
+
+    try {
+      await models.NotificationDeviceToken.upsert({
+        deviceToken,
+        deviceType,
+        userId
+      })
+
+      return successResponse()
+    } catch (e) {
+      req.logger.error(`Unable to register device token for userId: ${userId} on ${deviceType}`, e)
+      return errorResponseServerError(`Unable to register device token for userId: ${userId} on ${deviceType}, Error: ${e.message}`)
+    }
+  }))
+
+  /**
+   * Remove a device token from the device token table
+   * POST body contains {userId, deviceToken}
+   */
+  app.post('/push_notifications/device_token/deregister', handleResponse(async (req, res, next) => {
+    const { deviceToken, userId } = req.body
+    if (!deviceToken || !userId) return errorResponseBadRequest('Did not pass in a valid deviceToken or userId for device token registration')
+
+    let deleted = false
+    try {
+      const tokenObj = await models.NotificationDeviceToken.findOne({ where: {
+        userId,
```

Comment:
Since the primary key is `deviceToken` I don't think you need this, but I don't think it hurts.

Reply:
I think it may actually slow down the query since it has to satisfy both and it’ll have two separate query plans for it? Not sure but good point

Reply:
@jowlee yeah you're right this is unnecessary. i'll delete the userId. 

@raymondjacobson i believe if we left both conditions without an index it's actually faster with 2 filters  vs 1 because the query will need to do a seq scan anyway and it can filter on two filters vs 1 so it helps prune a huge part of the dataset. if there is an index, it's about the same performance because the query planner will pick the plan that runs the index scan before it scans the result of the index, so there should be no discernable performance hit.

Reply:
ty :)